### PR TITLE
HTML API: Return elements pushed and popped rather than tags read.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-61348

## Summary

Creates virtual nodes when pushing to and popping from the stack of open elements. It's these nodes that are returned by `next_tag()`, while subclassed methods intercept tag information, all within the HTML Processor.

## Splitting time!

 - [x] Add `get_current_depth()` to return the depth of the currently-matched element in the stack of open elements. This needs to account for `marker` and other not-yet-implemented items in the stack.
 - [x] Add `expects_closer()` or similar function to indicate if the currently-matched element needs or expects a closing element.

## Questions

 - What do we do with void tags?
   - There are tradeoffs in how to represent tags without a closing tag (be they void elements or self-closing elements within foreign content or special atomic elements like SCRIPT). We can represent a unique open and close tag/event for them, or only expose the opening.
     - In exposing an open and a close there's a very straightforward and universal model of sequential open and close events for every tag.
     - It's weird to have a closing tag for an IMG.
   - I think that the value in having a close tag is outweighed by the awkwardness of it. For situations where we're doing things like tracking depth, we should rely on an internal method like `get_depth()` from the HTML API itself instead of exporting that HTML nuance onto the caller.

## Related work

 - WordPress/wordpress-develop#5562
 - WordPress/gutenberg#46345

## Examples

| Convert to Markdown | Output |
|---|---|
| Screenshot 2024-04-03 at 6 15 43 PM | Screenshot 2024-04-03 at 6 17 06 PM |

| Normalize HTML | Output |
|---|---|
| Screenshot 2024-04-03 at 6 15 43 PM | Screenshot 2024-04-03 at 6 18 23 PM |

cc: @sirreal

### CODE REVIEW

**Comment by dmsnell** [comment]
@sirreal it looks like we're failing tests after seeking, which I believe might be because when jumping back to the start of the document, we wipe out the context node.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -424,28 +428,19 @@ public function next_tag( $query = null ) {
 					continue;
 				}
 
-				if ( ! parent::is_tag_closer() ) {
+				if ( ! parent::is_tag_closer() || $visit_closers ) {
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:466]: Should this also use `$this`?

```suggestion
if ( ! $this->is_tag_closer() || $visit_closers ) {
```

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:466]: Yes. Can you concoct a test case to fail for it? If not, don't worry - we can just fix it.

**Review by sirreal — COMMENTED** [review]
There's a regression here where this HTML:

```html
<?wp-bit hey?>
```

This is a `PI_NODE_LOOKALIKE`. Before, `get_tag()` here would be `wp-bit` - corresponding to the PI _target_. After this change `get_tag()` returns `#comment`.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -1327,6 +1387,10 @@ public function get_tag() {
 			return null;
 		}
 
+		if ( isset( $this->current_element ) ) {
+			return $this->current_element->token->node_name;
+		}
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:1444]: This is what breaks the `tag_name` for PI lookalikes. The node_name is `#comment`, but it's a special comment type that has more handling in the Tag Processor:

https://github.com/WordPress/wordpress-develop/blob/fbb5020a92735e1af1c1873ba6be000a8c19f0dd/src/wp-includes/html-api/class-wp-html-tag-processor.php#L2640-L2651

A potential fix is to only return here if we have a "non-special" node:

```suggestion
		if ( isset( $this->current_element ) && '#' !== $this->current_element->token->node_name[0] ) {
			return $this->current_element->token->node_name;
		}
```

**Comment by sirreal** [comment]
~The following HTML reaches a `P` tag at the wrong depth:~

**EDIT:** This is pre-existing behavior.

<details>

<summary>redacted</summary>

```html
<div></p>
```

```php
$p = WP_HTML_Processor::create_fragment('<div></p>');
$p->next_token();
$p->next_token();
var_dump( $p->get_tag(), $p->get_current_depth(), $p->get_breadcrumbs() );
```

```
php shell code:1:
string(1) "P"
php shell code:1:
int(3)
php shell code:1:
array(3) {
  [0] =>
  string(4) "HTML"
  [1] =>
  string(4) "BODY"
  [2] =>
  string(3) "DIV"
}
```

</details>

**Comment by dmsnell** [comment]
Merged in [[58304]](https://core.trac.wordpress.org/changeset/58304)
https://github.com/wordpress/wordpress-develop/commit/163c3fd11026e0a6f93a7f88ab32dd1679ad0f88

File: src/wp-includes/html-api/class-wp-html-open-elements.php [hunk 3]
```diff
@@ -51,6 +51,56 @@ class WP_HTML_Open_Elements {
...
+	 * Sets a pop handler that will be called when an item is popped off the stack of
+	 * open elements.
+	 *
+	 * The function will be called with the pushed item as its argument.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param Closure $handler The handler function.
+	 */
+	public function set_pop_handler( Closure $handler ) {
```
Comment (westonruter) [inline src/wp-includes/html-api/class-wp-html-open-elements.php:86]: Is there a reason this is specifically a `Closure` and isn't just more generally a `callable`?

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-open-elements.php:86]: Closures cannot be serialized or deserialized, meaning that there's no possible way to prep a database record with user input that sets something unexpected here.

**Pull request closed without merging**